zerver/lib/actions.py

```diff
@@ -1715,6 +1715,26 @@ def do_remove_reaction(user_profile: UserProfile, message: Message,
     reaction.delete()
     notify_reaction_update(user_profile, message, reaction, "remove")
 
+def do_remove_preview(message: Message, url: str) -> None:
+    preview_removed = PreviewRemoved(message=message, url=url)
+    try:
+        preview_removed.save()
+    except django.db.utils.IntegrityError:  # nocoverage
+        raise JsonableError(_("Preview already removed for this URL."))
+
+    # We render the message to populate the message.links_for_preview attribute
+    render_incoming_message(message,
+                            message.content,
+                            set(),
+                            message.sender.realm)
+    event_data = {
+        'message_id': message.id,
+        'message_content': message.content,
+        'message_realm_id': message.sender.realm_id,
+        'urls': message.links_for_preview}
```

Comment:
It would be nice if this event formatting code was shared with the other caller of the `embed_links` queue, so they can't end up diverging by accident in some future refactor.